File: specs/typespec/admin/personal-domain-blocklist.ts
import {
	type ValidationError,
	newValidationError,
} from "../common/common";

export interface BlockedPersonalDomain {
	domain: string;
	created_at: string;
}

export interface AdminAddBlockedDomainRequest {
	domain: string;
}

export interface AdminRemoveBlockedDomainRequest {
	domain: string;
}

export interface AdminListBlockedDomainsRequest {
	filter_domain_prefix?: string;
	pagination_key?: string;
	limit?: number;
}

export interface AdminListBlockedDomainsResponse {
	domains: BlockedPersonalDomain[];
	next_pagination_key?: string;
}

// Validation
export const ERR_DOMAIN_REQUIRED = "domain is required";
export const ERR_DOMAIN_TOO_LONG = "domain must be at most 253 characters";
export const ERR_DOMAIN_HAS_AT = "domain must not contain @";

export function validateAdminAddBlockedDomainRequest(
	request: AdminAddBlockedDomainRequest
): ValidationError[] {
	const errs: ValidationError[] = [];
	if (!request.domain || request.domain.trim() === "") {
		errs.push(newValidationError("domain", ERR_DOMAIN_REQUIRED));
	} else if (request.domain.length > 253) {
		errs.push(newValidationError("domain", ERR_DOMAIN_TOO_LONG));
	} else if (request.domain.includes("@")) {
		errs.push(newValidationError("domain", ERR_DOMAIN_HAS_AT));
	}
	return errs;
}

export function validateAdminRemoveBlockedDomainRequest(
	request: AdminRemoveBlockedDomainRequest
): ValidationError[] {
	const errs: ValidationError[] = [];
	if (!request.domain || request.domain.trim() === "") {
		errs.push(newValidationError("domain", ERR_DOMAIN_REQUIRED));
	}
	return errs;
}

export function validateAdminListBlockedDomainsRequest(
	request: AdminListBlockedDomainsRequest
): ValidationError[] {
	return [];
}

File: specs/typespec/admin/personal-domain-blocklist.go
package admin

import (
	"errors"
	"strings"

	"vetchium-api-server.typespec/common"
)

type BlockedPersonalDomain struct {
	Domain    string `json:"domain"`
	CreatedAt string `json:"created_at"`
}

type AdminAddBlockedDomainRequest struct {
	Domain string `json:"domain"`
}

type AdminRemoveBlockedDomainRequest struct {
	Domain string `json:"domain"`
}

type AdminListBlockedDomainsRequest struct {
	FilterDomainPrefix *string `json:"filter_domain_prefix,omitempty"`
	PaginationKey      *string `json:"pagination_key,omitempty"`
	Limit              *int32  `json:"limit,omitempty"`
}

type AdminListBlockedDomainsResponse struct {
	Domains           []BlockedPersonalDomain `json:"domains"`
	NextPaginationKey *string                 `json:"next_pagination_key,omitempty"`
}

var (
	ErrBlocklistDomainRequired = errors.New("domain is required")
	ErrBlocklistDomainTooLong  = errors.New("domain must be at most 253 characters")
	ErrBlocklistDomainHasAt    = errors.New("domain must not contain @")
)

func (r AdminAddBlockedDomainRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	domain := strings.TrimSpace(r.Domain)
	if domain == "" {
		errs = append(errs, common.NewValidationError("domain", ErrBlocklistDomainRequired))
	} else if len(domain) > 253 {
		errs = append(errs, common.NewValidationError("domain", ErrBlocklistDomainTooLong))
	} else if strings.Contains(domain, "@") {
		errs = append(errs, common.NewValidationError("domain", ErrBlocklistDomainHasAt))
	}
	return errs
}

func (r AdminRemoveBlockedDomainRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if strings.TrimSpace(r.Domain) == "" {
		errs = append(errs, common.NewValidationError("domain", ErrBlocklistDomainRequired))
	}
	return errs
}

func (r AdminListBlockedDomainsRequest) Validate() []common.ValidationError {
	return nil
}
